src/lib/datepicker/datepicker-input.ts

```diff
@@ -398,7 +398,104 @@ export class MatDatepickerInput<D> implements ControlValueAccessor, OnDestroy, V
    * @param obj The object to check.
    * @returns The given object if it is both a date instance and valid, otherwise null.
    */
-  private _getValidDateOrNull(obj: any): D | null {
+  protected _getValidDateOrNull(obj: any): D | null {
     return (this._dateAdapter.isDateInstance(obj) && this._dateAdapter.isValid(obj)) ? obj : null;
   }
 }
+
+@Directive({
+  selector: 'input[matDatepickerStart]',
+  providers: [
+    MAT_DATEPICKER_VALUE_ACCESSOR,
+    MAT_DATEPICKER_VALIDATORS,
+    {provide: MAT_INPUT_VALUE_ACCESSOR, useExisting: MatDatepickerInput},
+  ],
+  host: {
+    '[attr.aria-haspopup]': 'true',
+    '[attr.aria-owns]': '(_datepicker?.opened && _datepicker.id) || null',
+    '[attr.min]': 'min ? _dateAdapter.toIso8601(min) : null',
+    '[attr.max]': 'max ? _dateAdapter.toIso8601(max) : null',
+    '[disabled]': 'disabled',
+    '(input)': '_onInput($event.target.value)',
+    '(change)': '_onChange()',
+    '(blur)': '_onBlur()',
+    '(keydown)': '_onKeydown($event)',
+  },
+})
+export class MatDatepickerInputStart<D> extends MatDatepickerInput<D> {
+  @Input()
+  set matDatepicker(value: MatDatepicker<D>) {
+    if (!value) {
+      return;
+    }
+
+    this._datepicker = value;
+    this._datepickerSubscription.unsubscribe();
+
+    if (this._isSelectionInitialized) {
+      this._isSelectionInitialized = false;
+      this._selectionModel.ngOnDestroy();
+    }
+
+    this._selectionModel = this._datepicker._dateSelection;
+
+    this._formatValue(this._selectionModel.getFirstSelectedDate());
```

Comment:
Currently I implement the function as-is three times. In the base as well as the start and end directive. However the only thing that changes is `getSelection()` to `getFirstSelectedDate()` and `getLastSelectedDate()`. There is room for improvement to not have such duplicated code.

Reply:
Do you think it might help to have a base class (e.g. `MatDatepcikerInputBase`) that isn't an `@Directive` and then 3 `@Directive` classes that extend from it? My thinking is that the base class shouldn't know which type of selection model it's using. The single date input can assume it's a single date selection model, and the range start and end can assume it's a range selection.

Also, idea for how to abstract getting the date:

```
abstract class MatDatepickerInputBase<D> {
  abstract getSelectedDate(): D | null;
  ...
}

class MatDatepickerInput<D> extends MatDatepickerInputBase<D> {
  getSelectedDate(): D | null { return this.selectionModel.getSelection(); }
  ...
}

class MatDatepickerInputStart<D> extends MatDatepickerInputBase<D> {
  getSelectedDate(): D | null { return this.selectionModel.getFirstSelectedDate(); }
  ...
}
```

Reply:
This would be nice. Right now it always says `getFirstSelectedDate` which seems confusing when it's  a single date.